File: python/nymrel_proof_ledger/signer.py
"""HMAC-SHA256 and audited-library Ed25519 signing primitives."""

from __future__ import annotations

import hashlib
import hmac
import re
import secrets
from datetime import UTC, datetime
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from .canonical import CanonicalizationProfile, canonicalize

HEX_32_BYTES = re.compile(r"^[0-9a-fA-F]{64}$")
HMAC_SIGNATURE = re.compile(r"^[0-9a-fA-F]{64}$")
ED25519_SIGNATURE = re.compile(r"^[0-9a-fA-F]{128}$")


def _raw_key(material: str, label: str) -> bytes:
    if not isinstance(material, str) or HEX_32_BYTES.fullmatch(material) is None:
        raise ValueError(
            f"{label} must be exactly 32 bytes encoded as 64 hexadecimal characters"
        )
    return bytes.fromhex(material)


class ProofSigner:
    @staticmethod
    def generate_secret_key() -> str:
        return secrets.token_hex(32)

    @staticmethod
    def generate_key_pair() -> dict[str, str]:
        private_key = Ed25519PrivateKey.generate()
        public_key = private_key.public_key()
        return {
            "privateKey": private_key.private_bytes(
                serialization.Encoding.Raw,
                serialization.PrivateFormat.Raw,
                serialization.NoEncryption(),
            ).hex(),
            "publicKey": public_key.public_bytes(
                serialization.Encoding.Raw,
                serialization.PublicFormat.Raw,
            ).hex(),
            "algorithm": "Ed25519",
            "encoding": "raw-hex",
        }

    @staticmethod
    def sign_payload(
        payload: Any,
        private_key_or_secret: str,
        algorithm: str = "HMAC-SHA256",
        canonicalization_profile: CanonicalizationProfile = "rfc8785",
    ) -> str:
        if not isinstance(private_key_or_secret, str) or not private_key_or_secret:
            raise ValueError('Signing key must be a non-empty string')
        canonicalize(private_key_or_secret)
        canonical = (
            payload
            if isinstance(payload, str)
            else canonicalize(payload, canonicalization_profile)
        )
        message = canonical.encode("utf-8")
        if algorithm == "HMAC-SHA256":
            return hmac.new(
                private_key_or_secret.encode("utf-8"), message, hashlib.sha256
            ).hexdigest()
        if algorithm == "Ed25519":
            private_key = Ed25519PrivateKey.from_private_bytes(
                _raw_key(private_key_or_secret, "Ed25519 private key")
            )
            return private_key.sign(message).hex()
        raise ValueError(f"Unsupported signature algorithm: {algorithm}")

    @staticmethod
    def verify_signature(
        payload: Any,
        signature_hex: str,
        public_key_or_secret: str,
        algorithm: str = "HMAC-SHA256",
        canonicalization_profile: CanonicalizationProfile = "rfc8785",
    ) -> bool:
        try:
            if not isinstance(public_key_or_secret, str) or not public_key_or_secret:
                return False
            canonicalize(public_key_or_secret)
            canonical = (
                payload
                if isinstance(payload, str)
                else canonicalize(payload, canonicalization_profile)
            )
            message = canonical.encode("utf-8")
            if algorithm == "HMAC-SHA256":
                if HMAC_SIGNATURE.fullmatch(signature_hex) is None:
                    return False
                expected = hmac.new(
                    public_key_or_secret.encode("utf-8"), message, hashlib.sha256
                ).hexdigest()
                return hmac.compare_digest(signature_hex.lower(), expected)
            if algorithm == "Ed25519":
                if ED25519_SIGNATURE.fullmatch(signature_hex) is None:
                    return False
                public_key = Ed25519PublicKey.from_public_bytes(
                    _raw_key(public_key_or_secret, "Ed25519 public key")
                )
                public_key.verify(bytes.fromhex(signature_hex), message)
                return True
            return False
        except (InvalidSignature, TypeError, ValueError):
            return False

    @staticmethod
    def create_signature_record(
        payload: Any,
        private_key_or_secret: str,
        signer_identity: str,
        key_id: str,
        algorithm: str = "HMAC-SHA256",
        canonicalization_profile: CanonicalizationProfile = "rfc8785",
    ) -> dict[str, str]:
        return {
            "algorithm": algorithm,
            "keyId": key_id,
            "signerIdentity": signer_identity,
            "value": ProofSigner.sign_payload(
                payload,
                private_key_or_secret,
                algorithm,
                canonicalization_profile,
            ),
            "timestamp": datetime.now(UTC).isoformat().replace("+00:00", "Z"),
        }

File: src/core/signer.ts
/** HMAC-SHA256 and Ed25519 signing over canonical payloads. */

import {
  createHmac,
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  randomBytes,
  sign as cryptoSign,
  timingSafeEqual,
  verify as cryptoVerify,
  type KeyObject,
} from 'node:crypto';
import { canonicalize, type CanonicalizationProfile } from './canonical.js';

export type SignatureAlgorithm = 'HMAC-SHA256' | 'Ed25519';

export interface KeyPair {
  publicKey: string;
  privateKey: string;
  algorithm: 'Ed25519';
  encoding: 'raw-hex';
}

export interface SignatureRecord {
  algorithm: SignatureAlgorithm;
  keyId: string;
  signerIdentity: string;
  value: string;
  timestamp: string;
}

const HEX_32_BYTES = /^[0-9a-fA-F]{64}$/;
const HMAC_SIGNATURE = /^[0-9a-fA-F]{64}$/;
const ED25519_SIGNATURE = /^[0-9a-fA-F]{128}$/;
const ED25519_PKCS8_PREFIX = Buffer.from('302e020100300506032b657004220420', 'hex');
const ED25519_SPKI_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');

function privateEd25519Key(material: string): KeyObject {
  if (HEX_32_BYTES.test(material)) {
    return createPrivateKey({
      key: Buffer.concat([ED25519_PKCS8_PREFIX, Buffer.from(material, 'hex')]),
      format: 'der',
      type: 'pkcs8',
    });
  }
  const key = createPrivateKey(material);
  if (key.asymmetricKeyType !== 'ed25519') throw new TypeError('Private key is not Ed25519');
  return key;
}

function publicEd25519Key(material: string): KeyObject {
  if (HEX_32_BYTES.test(material)) {
    return createPublicKey({
      key: Buffer.concat([ED25519_SPKI_PREFIX, Buffer.from(material, 'hex')]),
      format: 'der',
      type: 'spki',
    });
  }
  const key = createPublicKey(material);
  if (key.asymmetricKeyType !== 'ed25519') throw new TypeError('Public key is not Ed25519');
  return key;
}

export class ProofSigner {
  public static generateSecretKey(): string {
    return randomBytes(32).toString('hex');
  }

  public static generateKeyPair(): KeyPair {
    const pair = generateKeyPairSync('ed25519');
    const privateDer = pair.privateKey.export({ format: 'der', type: 'pkcs8' });
    const publicDer = pair.publicKey.export({ format: 'der', type: 'spki' });
    return {
      privateKey: privateDer.subarray(privateDer.length - 32).toString('hex'),
      publicKey: publicDer.subarray(publicDer.length - 32).toString('hex'),
      algorithm: 'Ed25519',
      encoding: 'raw-hex',
    };
  }

  public static signPayload(
    payload: unknown,
    privateKeyOrSecret: string,
    algorithm: SignatureAlgorithm = 'HMAC-SHA256',
    canonicalizationProfile: CanonicalizationProfile = 'rfc8785'
  ): string {
    const canonicalPayload = typeof payload === 'string'
      ? payload
      : canonicalize(payload, canonicalizationProfile);
    const payloadBuffer = Buffer.from(canonicalPayload, 'utf8');
    if (algorithm === 'HMAC-SHA256') {
      return createHmac('sha256', privateKeyOrSecret).update(payloadBuffer).digest('hex');
    }
    if (algorithm === 'Ed25519') {
      return cryptoSign(null, payloadBuffer, privateEd25519Key(privateKeyOrSecret)).toString('hex');
    }
    throw new Error(`Unsupported signature algorithm: ${String(algorithm)}`);
  }

  public static verifySignature(
    payload: unknown,
    signatureHex: string,
    publicKeyOrSecret: string,
    algorithm: SignatureAlgorithm = 'HMAC-SHA256',
    canonicalizationProfile: CanonicalizationProfile = 'rfc8785'
  ): boolean {
    try {
      const canonicalPayload = typeof payload === 'string'
        ? payload
        : canonicalize(payload, canonicalizationProfile);
      const payloadBuffer = Buffer.from(canonicalPayload, 'utf8');
      if (algorithm === 'HMAC-SHA256') {
        if (!HMAC_SIGNATURE.test(signatureHex)) return false;
        const expected = createHmac('sha256', publicKeyOrSecret).update(payloadBuffer).digest();
        return timingSafeEqual(Buffer.from(signatureHex, 'hex'), expected);
      }
      if (algorithm === 'Ed25519') {
        if (!ED25519_SIGNATURE.test(signatureHex)) return false;
        return cryptoVerify(
          null,
          payloadBuffer,
          publicEd25519Key(publicKeyOrSecret),
          Buffer.from(signatureHex, 'hex')
        );
      }
      return false;
    } catch {
      return false;
    }
  }

  public static createSignatureRecord(
    payload: unknown,
    privateKeyOrSecret: string,
    signerIdentity: string,
    keyId: string,
    algorithm: SignatureAlgorithm = 'HMAC-SHA256',
    canonicalizationProfile: CanonicalizationProfile = 'rfc8785'
  ): SignatureRecord {
    return {
      algorithm,
      keyId,
      signerIdentity,
      value: ProofSigner.signPayload(payload, privateKeyOrSecret, algorithm, canonicalizationProfile),
      timestamp: new Date().toISOString(),
    };
  }
}
